### ISSUE TITLE
Same position services

### ISSUE THREAD

**Issue by braktar** [VROOM-Project/vroom#79]
Hi,

By sending a jobs at the same place than a depot I expect to find it next to the start point, which is not the case. This job is not returned at all.

Problem send : 
`
{
    "vehicles": [
    {
        "id": 0,
        "start_index": 0
    }],
    "jobs": [
    {
        "id": 0,
        "location_index": 0
    },
    {
        "id": 1,
        "location_index": 1
    }],
    "matrix": [
        [0, 1],
        [1, 0]
    ]
}
`

And the associated result
`
{
    "code": 0,
    "summary":
    {
        "cost": 1,
        "computing_times":
        {
            "loading": 0,
            "solving": 0
        }
    },
    "routes": [
    {
        "vehicle": 0,
        "cost": 1,
        "steps": [
        {
            "type": "start"
        },
        {
            "type": "job",
            "job": 1
        }]
    }]
}
`

I agree there is nothing to perform, but it the case of the CVRP, it could become an issue.
Every send jpb should be in the solution even if it's useless from the resolution point of view

**Comment by jcoupey** [comment]
This is not a solving issue but rather related to parsing, as it only occurs when using `location_index` with a custom matrix. This input in the same situation without matrix

```
{"jobs":[{"location":[5,47],"id":0},{"location":[6,48],"id":1}],"vehicles":[{"start":[5,47],"id":0}]}
```

 does work as expected.

**Comment by braktar** [comment]
Thanks,

I've done quite the same trick by my side, extending the matrix by duplicating the row and column from the jobs with the same location_index.

I could probably filter the services by my side as long as the clustering heuristic CVRP is not released :)

**Comment by jcoupey** [comment]
The current status is that `location_index` should be unique to keep out of trouble. As you already noticed, adding rows and columns to extend the matrix and avoid duplicate `location_index` works fine, even if this is just a workaround.

What happens is that indices are stored in a `std::set` in `input`, so jobs with a duplicate `location_index` are simply not passed on to `tsp` for solving. I didn't see that coming when merging #59. There is more to it than just making sure we keep duplicates, as part of the code relies on a mapping between the index and the actual job which does not make sense anymore if we allow duplicate index.

**Comment by jcoupey** [comment]
As far as I can tell, the changes currently in branch `fix/duplicate-index` should fix the problem for instances with duplicate `location_index` values. Hopefully nothing is changed for other instances, be it with unique `location_index` or using OSRM.

Note: the matrix with extended row and columns is actually built in the background while solving so this is just a convenient way to allow users to input a minimal matrix with locations sharing the same index.

@braktar do you have a few "serious" instances on which you could test that branch?

**Comment by braktar** [comment]
Thanks,

It could be really valuable on large instances. I will test it with various real cases

**Comment by braktar** [comment]
The current fix/duplicate-index branch only return SegFault even for problem without duplicated indices which works perfectly on the current master.

Problem send : 
`{"vehicles":[{"id":0,"start_index":4}],"jobs":[{"id":0,"location_index":0},{"id":1,"location_index":1},{"id":2,"location_index":2},{"id":3,"location_index":3}],"matrix":[[0,655,1948,5231,2971],[603,0,1692,4977,2715],[1861,1636,0,6143,1532],[5184,4951,6221,0,7244],[2982,2758,1652,7264,0]]}`

**Comment by jcoupey** [comment]
@braktar my guess is that you're hitting the problem currently tracked at #78, and not all dependent files were recompiled. Can you try to `make clean` then `make` to check a fresh build?

**Comment by braktar** [comment]
Ok, thanks. It actually passes the tests cases by my side. I will now try real cases

**Comment by braktar** [comment]
Seems to be correct !
Request : 
`{"vehicles":[{"id":0,"start_index":21,"end_index":21}],"jobs":[{"id":0,"location_index":0},{"id":1,"location_index":0},{"id":2,"location_index":1},{"id":3,"location_index":2},{"id":4,"location_index":3},{"id":5,"location_index":4},{"id":6,"location_index":4},{"id":7,"location_index":4},{"id":8,"location_index":5},{"id":9,"location_index":5},{"id":10,"location_index":6},{"id":11,"location_index":6},{"id":12,"location_index":7},{"id":13,"location_index":8},{"id":14,"location_index":8},{"id":15,"location_index":9},{"id":16,"location_index":10},{"id":17,"location_index":10},{"id":18,"location_index":10},{"id":19,"location_index":11},{"id":20,"location_index":11},{"id":21,"location_index":12},{"id":22,"location_index":13},{"id":23,"location_index":14},{"id":24,"location_index":15},{"id":25,"location_index":15},{"id":26,"location_index":16},{"id":27,"location_index":17},{"id":28,"location_index":17},{"id":29,"location_index":18},{"id":30,"location_index":18},{"id":31,"location_index":18},{"id":32,"location_index":19},{"id":33,"location_index":20},{"id":34,"location_index":20},{"id":35,"location_index":20}],"matrix":[[0,22851,13751,14458,11201,21014,10155,18334,19518,17912,11597,16523,15527,14153,21114,19207,25268,13178,14473,23615,24740,14737],[22519,0,14342,13679,23172,15464,16723,7217,12254,6029,12020,7044,15567,17702,11641,7881,9109,11733,8670,12311,7746,9353],[13787,14358,0,12207,18139,18513,14983,9840,15283,9416,9302,10725,10697,17272,16671,10008,17625,11896,10352,19171,15644,10131],[14437,14017,12143,0,13559,10747,6930,10441,9830,9813,3647,7549,9504,9139,11998,13387,16151,3962,5498,14498,15769,5848],[11206,23622,18098,13533,0,20366,9777,20125,19448,19497,12737,17156,17449,14133,21475,21499,25629,12783,15171,23975,25373,15515],[20910,15686,18450,10702,20314,0,13320,15190,4578,14345,10922,9453,16600,11594,9868,18188,12571,7920,8741,7677,16834,9441],[10084,17070,15015,6892,9704,13349,0,13535,12431,12907,6520,10604,12377,7199,14923,16433,19078,5511,8594,17423,18821,8944],[18416,7199,9828,10486,20033,15333,13582,0,11618,2746,8623,6424,11053,14704,12525,6828,11695,8735,6992,14896,10535,6970],[19520,12596,15234,9802,19415,4581,12420,11649,0,10450,9020,5558,13938,11520,5949,14972,10073,7020,5478,7451,14336,6350],[17973,6021,9384,9912,19458,14136,13008,2752,10240,0,8047,5029,10608,14129,11148,7148,9953,8160,6417,13155,8908,6394],[11822,12497,9528,3715,12758,11045,6492,8699,9026,8070,0,6030,6889,8782,10623,10945,14776,4259,3978,13123,14249,4243],[16390,7268,10526,7548,17043,9417,10593,6512,5521,5120,5888,0,10415,11571,6335,10265,9744,5601,2535,9571,9021,3220],[15530,15523,10699,9569,17497,16578,12345,11006,13757,10589,6663,10604,0,14633,15303,12116,18670,9887,8752,17803,17498,8756],[14134,18128,17294,9149,14095,11588,7194,14783,11509,14155,8800,11662,14655,0,15981,17729,20135,6121,9651,17463,19879,9995],[21128,11849,16497,12043,21423,9890,14972,12482,5964,11262,10629,6243,15404,15951,0,15417,10294,9983,7053,8997,13615,7967],[19289,7834,10052,13412,21549,18277,16473,6816,15048,7156,10793,10297,12186,17630,15165,0,12633,11661,10117,15835,8819,9895],[25155,9303,17576,16137,25516,12520,19067,11733,10044,10031,14656,9681,18753,20045,10231,12871,0,14077,11148,9274,9385,11990],[13066,12121,11984,4014,12655,7952,5474,8776,7034,8147,4233,5653,10076,6044,9975,11722,14129,0,3641,12475,13874,3986],[14287,9059,10248,5445,14941,8777,8489,6987,5435,6358,3785,2590,8712,9468,7039,9986,11193,3498,0,9540,10812,1453],[23692,12509,19085,14607,23987,7663,17536,14938,7442,13236,13193,9691,17968,17424,9037,16076,9323,12547,9618,0,13511,10532],[24377,7832,15682,15538,25031,16860,18581,10582,14113,8976,13879,8903,17582,19560,13698,8767,10175,13591,10528,13512,0,11212],[14695,9719,9991,5853,15397,9367,8945,6782,6249,6154,4193,3250,8804,10081,7853,9729,12007,4110,1367,10353,11471,0]]}
`

Answer:
`
{"code":0,"summary":{"cost":155479,"computing_times":{"loading":0,"solving":1}},"routes":[{"vehicle":0,"cost":155479,"steps":[{"type":"start"},{"type":"job","job":31},{"type":"job","job":30},{"type":"job","job":29},{"type":"job","job":19},{"type":"job","job":20},{"type":"job","job":23},{"type":"job","job":14},{"type":"job","job":13},{"type":"job","job":8},{"type":"job","job":9},{"type":"job","job":32},{"type":"job","job":26},{"type":"job","job":33},{"type":"job","job":35},{"type":"job","job":34},{"type":"job","job":2},{"type":"job","job":25},{"type":"job","job":24},{"type":"job","job":12},{"type":"job","job":15},{"type":"job","job":28},{"type":"job","job":27},{"type":"job","job":22},{"type":"job","job":10},{"type":"job","job":11},{"type":"job","job":6},{"type":"job","job":5},{"type":"job","job":7},{"type":"job","job":1},{"type":"job","job":0},{"type":"job","job":3},{"type":"job","job":21},{"type":"job","job":16},{"type":"job","job":17},{"type":"job","job":18},{"type":"job","job":4},{"type":"end"}]}]}
`

36 jobs in input and ouput

**Comment by jcoupey** [comment]
I just merged #81 so this should be fixed in `master`.
